File: src/main/java/net/gtaun/shoebill/common/dialog/AbstractDialog.java
/**
 * Copyright (C) 2012-2014 MK124
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.common.dialog;

import java.util.function.Supplier;

import net.gtaun.shoebill.SampObjectManager;
import net.gtaun.shoebill.constant.DialogStyle;
import net.gtaun.shoebill.event.dialog.DialogCloseEvent;
import net.gtaun.shoebill.event.dialog.DialogCloseEvent.DialogCloseType;
import net.gtaun.shoebill.event.dialog.DialogResponseEvent;
import net.gtaun.shoebill.object.DialogId;
import net.gtaun.shoebill.object.Player;
import net.gtaun.util.event.Attentions;
import net.gtaun.util.event.EventManager;
import net.gtaun.util.event.EventManagerNode;
import net.gtaun.util.event.HandlerPriority;

/**
 * 
 * @author MK124
 */
public abstract class AbstractDialog
{
	@SuppressWarnings("unchecked")
	public static abstract class AbstractDialogBuilder
	<DialogType extends AbstractDialog, DialogBuilderType extends AbstractDialogBuilder<DialogType, DialogBuilderType>>
	{
		protected final DialogType dialog;
		
		
		protected AbstractDialogBuilder(DialogType dialog)
		{
			this.dialog = dialog;
		}
		
		public DialogBuilderType parentDialog(AbstractDialog parent)
		{
			dialog.setParentDialog(parent);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType caption(String caption)
		{
			dialog.setCaption(caption);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType caption(DialogTextSupplier captionSupplier)
		{
			dialog.setCaption(captionSupplier);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType buttonOk(String buttonOk)
		{
			dialog.setButtonOk(buttonOk);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType buttonOk(DialogTextSupplier buttonOkSupplier)
		{
			dialog.setButtonOk(buttonOkSupplier);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType buttonCancel(String buttonCancel)
		{
			dialog.setButtonCancel(buttonCancel);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType buttonCancel(DialogTextSupplier buttonCancelSupplier)
		{
			dialog.setButtonCancel(buttonCancelSupplier);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType onClickCancel(DialogHandler handler)
		{
			dialog.setClickCancelHandler(handler);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType onShow(DialogHandler handler)
		{
			dialog.setShowHandler(handler);
			return (DialogBuilderType) this;
		}
		
		public DialogBuilderType onClose(DialogCloseHandler handler)
		{
			dialog.setCloseHandler(handler);
			return (DialogBuilderType) this;
		}
		
		public DialogType build()
		{
			return dialog;
		}
	}
	
	
	@FunctionalInterface
	public interface DialogCloseHandler
	{
		void onClose(AbstractDialog dialog, DialogCloseType type);
	}
	
	
	protected final Player player;
	protected final EventManagerNode eventManagerNode;
	private final EventManagerNode eventManagerInternal;
	
	private final DialogId dialogId;
	private final DialogStyle style;

	private AbstractDialog parentDialog;
	
	private DialogTextSupplier captionSupplier = (d) -> "None";
	private DialogTextSupplier buttonOkSupplier = (d) -> "OK";
	private DialogTextSupplier buttonCancelSupplier = (d) -> "Cancel";

	private DialogHandler showHandler = null;
	private DialogCloseHandler closeHandler = null;
	private DialogHandler clickCancelHandler = null;
	
	
	AbstractDialog(DialogStyle style, Player player, EventManager rootEventManager)
	{
		this.style = style;
		this.player = player;
		
		eventManagerInternal = rootEventManager.createChildNode();
		eventManagerNode = rootEventManager.createChildNode();
		
		dialogId = SampObjectManager.get().createDialogId();
	}
	
	@Override
	protected void finalize() throws Throwable
	{
		destroy();
		super.finalize();
	}
	
	protected void destroy()
	{
		eventManagerInternal.destroy();
		eventManagerNode.destroy();
	}
	
	public Player getPlayer()
	{
		return player;
	}
	
	public DialogStyle getStyle()
	{
		return style;
	}

	public AbstractDialog getParentDialog()
	{
		return parentDialog;
	}
	
	public void setParentDialog(AbstractDialog parentDialog)
	{
		this.parentDialog = parentDialog;
	}

	public void showParentDialog()
	{
		eventManagerNode.cancelAll();
		if (parentDialog != null) parentDialog.show();
	}
	
	public void setCaption(String caption)
	{
		captionSupplier = (d) -> caption;
	}
	
	public void setCaption(Supplier<String> captionSupplier)
	{
		this.captionSupplier = (d) -> captionSupplier.get();
	}
	
	public void setCaption(DialogTextSupplier captionSupplier)
	{
		this.captionSupplier = captionSupplier;
	}
	
	public String getCaption()
	{
		return captionSupplier.get(this);
	}
	
	public void setButtonOk(String buttonOk)
	{
		buttonOkSupplier = (d) -> buttonOk;
	}

	public void setButtonOk(Supplier<String> buttonOkSupplier)
	{
		this.buttonOkSupplier = (d) -> buttonOkSupplier.get();
	}

	public void setButtonOk(DialogTextSupplier buttonOkSupplier)
	{
		this.buttonOkSupplier = buttonOkSupplier;
	}
	
	public void setButtonCancel(String buttonCancel)
	{
		buttonCancelSupplier = (d) -> buttonCancel;
	}
	
	public void setButtonCancel(Supplier<String> buttonCancelSupplier)
	{
		this.buttonCancelSupplier = (d) -> buttonCancelSupplier.get();
	}
	
	public void setButtonCancel(DialogTextSupplier buttonCancelSupplier)
	{
		this.buttonCancelSupplier = buttonCancelSupplier;
	}
	
	public void setClickCancelHandler(DialogHandler onClickCancelHandler)
	{
		this.clickCancelHandler = onClickCancelHandler;
	}
	
	public void setShowHandler(DialogHandler showHandler)
	{
		this.showHandler = showHandler;
	}
	
	public void setCloseHandler(DialogCloseHandler closeHandler)
	{
		this.closeHandler = closeHandler;
	}
	
	protected void show(String text)
	{
		onShow();
		
		eventManagerInternal.cancelAll();
		eventManagerInternal.registerHandler(DialogResponseEvent.class, HandlerPriority.NORMAL, Attentions.create().object(dialogId), (e) ->
		{
			if (e.getDialogResponse() == 1)
			{
				onClickOk(e);
			}
			else
			{
				onClickCancel();
			}
			
			onClose(DialogCloseType.RESPOND);
		});

		eventManagerInternal.registerHandler(DialogCloseEvent.class, HandlerPriority.NORMAL, Attentions.create().object(dialogId), (e) ->
		{
			eventManagerInternal.cancelAll();
			onClose(e.getType());
		});
		
		player.showDialog(dialogId, style, captionSupplier.get(this), text, buttonOkSupplier.get(this), buttonCancelSupplier.get(this));
	}
	
	public abstract void show();
	
	protected void onShow()
	{
		if (showHandler != null) showHandler.handle(this);
	}

	protected void onClose(DialogCloseType type)
	{
		if (closeHandler != null) closeHandler.onClose(this, type);
	}

	void onClickOk(DialogResponseEvent event)
	{
		
	}
	
	protected void onClickCancel()
	{
		if (clickCancelHandler != null) clickCancelHandler.handle(this);
	}
}

File: src/main/java/net/gtaun/shoebill/common/dialog/AbstractDialog.kt
/**
 * Copyright (C) 2012-2016 MK124

 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.gtaun.shoebill.common.dialog

import net.gtaun.shoebill.common.AllOpen
import net.gtaun.shoebill.constant.DialogStyle
import net.gtaun.shoebill.entities.DialogId
import net.gtaun.shoebill.entities.Player
import net.gtaun.shoebill.event.dialog.DialogCloseEvent
import net.gtaun.shoebill.event.dialog.DialogCloseEvent.DialogCloseType
import net.gtaun.shoebill.event.dialog.DialogResponseEvent
import net.gtaun.util.event.Attentions
import net.gtaun.util.event.EventManager
import net.gtaun.util.event.EventManagerNode
import net.gtaun.util.event.HandlerPriority


/**
 * @author MK124
 * @author Marvin Haschker
 */
@AllOpen
abstract class AbstractDialog(protected var style: DialogStyle, parentEventManager: EventManager) {

    @Suppress("UNCHECKED_CAST")
    @AllOpen
    class Builder<T : AbstractDialog, B : Builder<T, B>> {

        @Suppress("ProtectedInFinal")
        lateinit protected var dialog: T

        fun init(init: B.(B) -> Unit): B {
            init(this as B, this)
            return this
        }

        fun caption(caption: String) = caption { caption }
        fun buttonOk(buttonOk: String) = buttonOk { buttonOk }
        fun buttonCancel(buttonCancel: String) = buttonCancel { buttonCancel }

        fun captionSupplier(supplier: DialogTextSupplier): B {
            dialog.captionSupplier = supplier
            return this as B
        }

        fun buttonOkSupplier(supplier: DialogTextSupplier): B {
            dialog.buttonOkSupplier = supplier
            return this as B
        }

        fun buttonCancelSupplier(supplier: DialogTextSupplier): B {
            dialog.buttonCancelSupplier = supplier
            return this as B
        }

        fun onShow(handler: DialogHandler) = onShow { handler }
        fun onClose(handler: DialogCloseHandler) = onClose { handler }
        fun onCancel(handler: DialogHandler) = onCancel { handler }
        fun parentDialog(parentDialog: AbstractDialog) = parentDialog { parentDialog }

        fun caption(init: B.() -> String): B {
            dialog.caption = init(this as B)
            return this
        }

        fun captionSupplier(init: B.(AbstractDialog) -> String): B =
                captionSupplier(DialogTextSupplier { init(this as B, it) })

        fun buttonOk(init: B.() -> String): B {
            dialog.buttonOk = init(this as B)
            return this
        }

        fun buttonOkSupplier(init: B.(AbstractDialog) -> String): B =
                buttonOkSupplier(DialogTextSupplier { init(this as B, it) })

        fun buttonCancel(init: B.() -> String): B {
            dialog.buttonCancel = init(this as B)
            return this
        }

        fun buttonCancelSupplier(init: B.(AbstractDialog) -> String): B =
                buttonCancelSupplier(DialogTextSupplier { init(this as B, it) })

        fun onShow(init: B.() -> DialogHandler): B {
            dialog.showHandler = init(this as B)
            return this
        }

        fun onClose(init: B.() -> DialogCloseHandler): B {
            dialog.closeHandler = init(this as B)
            return this
        }

        fun onCancel(init: B.() -> DialogHandler): B {
            dialog.clickCancelHandler = init(this as B)
            return this
        }

        fun parentDialog(init: B.() -> AbstractDialog): B {
            dialog.parentDialog = init(this as B)
            return this
        }

        fun build(): T = dialog
    }

    @FunctionalInterface
    interface DialogCloseHandler {
        fun onClose(dialog: AbstractDialog, player: Player, type: DialogCloseType)
    }

    private val eventManagerInternal: EventManagerNode = parentEventManager.createChildNode()
    val dialogId: DialogId = DialogId.create()

    init {
        eventManagerInternal.registerHandler(DialogResponseEvent::class, { e ->
            onClose(DialogCloseType.RESPOND, e.player)

            if (e.dialogResponse == 1) {
                onClickOk(e)
            } else {
                onClickCancel(e.player)
            }

        }, HandlerPriority.NORMAL, Attentions.create().obj(dialogId))

        eventManagerInternal.registerHandler(DialogCloseEvent::class, { e ->
            if (e.type != DialogCloseType.RESPOND) {
                onClose(e.type, e.player)
            }
        }, HandlerPriority.NORMAL, Attentions.create().obj(dialogId))
    }

    var parentDialog: AbstractDialog? = null

    var captionSupplier = DialogTextSupplier { "None" }
    var buttonOkSupplier = DialogTextSupplier { "OK" }
    var buttonCancelSupplier = DialogTextSupplier { "Cancel" }

    var showHandler: DialogHandler? = null
    var closeHandler: DialogCloseHandler? = null
    var clickCancelHandler: DialogHandler? = null

    @Throws(Throwable::class)
    protected fun finalize() {
        destroy()
    }

    protected fun destroy() {
        eventManagerInternal.destroy()
        dialogId.destroy()
    }

    fun showParentDialog(player: Player) = parentDialog?.show(player)

    fun setCaption(captionSupplier: DialogTextSupplier) {
        this.captionSupplier = captionSupplier
    }

    var caption: String
        get() = captionSupplier[this]
        set(caption) {
            captionSupplier = DialogTextSupplier { caption }
        }

    fun setButtonOk(buttonOkSupplier: DialogTextSupplier) {
        this.buttonOkSupplier = buttonOkSupplier
    }

    fun setButtonCancel(buttonCancelSupplier: DialogTextSupplier) {
        this.buttonCancelSupplier = buttonCancelSupplier
    }

    var buttonOk: String
        get() = buttonOkSupplier[this]
        set(buttonOk) {
            buttonOkSupplier = DialogTextSupplier { buttonOk }
        }

    var buttonCancel: String
        get() = buttonCancelSupplier[this]
        set(buttonCancel) {
            buttonCancelSupplier = DialogTextSupplier { buttonCancel }
        }

    fun show(player: Player, text: String) {
        onShow(player)
        player.showDialog(dialogId, style, captionSupplier[this], text,
                buttonOkSupplier[this], buttonCancelSupplier[this])
    }

    abstract fun show(player: Player)

    fun onShow(player: Player) = showHandler?.handle(this, player)

    fun onClickOk(event: DialogResponseEvent) {}

    fun onClose(type: DialogCloseType, player: Player) = closeHandler?.onClose(this, player, type)
    fun onClickCancel(player: Player) = clickCancelHandler?.handle(this, player)

    companion object {
        fun DialogCloseHandler(handler: (AbstractDialog, Player, DialogCloseType) -> Unit) = object : DialogCloseHandler {
            override fun onClose(dialog: AbstractDialog, player: Player, type: DialogCloseType) {
                handler(dialog, player, type)
            }
        }
    }
}
